File: dsa-problems/GraphsBasics/solutions/python/GRB-002-dfs-connected-components.py
def components(n: int, adj: list[list[int]]) -> list[int]:
    comp = [0] * n
    count = 0
    stack = []

    for i in range(n):
        if comp[i] != 0:
            continue
        count += 1
        stack.append(i)
        comp[i] = count
        while stack:
            u = stack.pop()
            for v in adj[u]:
                if comp[v] == 0:
                    comp[v] = count
                    stack.append(v)
    return comp

def main():
    import sys
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    n = int(next(it))
    m = int(next(it))
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u = int(next(it))
        v = int(next(it))
        adj[u].append(v)
        adj[v].append(u)
    comp = components(n, adj)
    max_comp = max(comp) if comp else 0
    sys.stdout.write(str(max_comp))

if __name__ == "__main__":
    main()

File: dsa-problems/GraphsBasics/solutions/javascript/GRB-002-dfs-connected-components.js
const readline = require("readline");

class Solution {
  components(n, adj) {
    const comp = new Int32Array(n).fill(0);
    let count = 0;
    const stack = [];

    for (let i = 0; i < n; i++) {
      if (comp[i] !== 0) continue;
      count++;
      stack.length = 0;
      stack.push(i);
      comp[i] = count;
      while (stack.length > 0) {
        const u = stack.pop();
        for (const v of adj[u]) {
          if (comp[v] === 0) {
            comp[v] = count;
            stack.push(v);
          }
        }
      }
    }
    return comp;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const m = parseInt(data[idx++], 10);
  const adj = Array.from({ length: n }, () => []);
  for (let i = 0; i < m; i++) {
    const u = parseInt(data[idx++], 10);
    const v = parseInt(data[idx++], 10);
    adj[u].push(v);
    adj[v].push(u);
  }

  const solution = new Solution();
  const comp = solution.components(n, adj);
  let maxComp = 0;
  for (let i = 0; i < n; i++) {
    if (comp[i] > maxComp) maxComp = comp[i];
  }
  console.log(maxComp.toString());
});
